tsfresh/feature_extraction/data.py

```diff
@@ -0,0 +1,351 @@
+        kinds = self.value_columns[column_offset:]
+        i = 0
+
+        for group_name, group in itertools.islice(self.df_grouped, group_offset, None):
+            for kind in kinds:
+                i += 1
+                if length is not None and i > length:
+                    return
+                yield Timeseries(group_name, kind, group[kind])
+            kinds = self.value_columns
+
+
+class LongTsFrameAdapter(TsData):
+
+    def __init__(self, df, column_id, column_kind, column_value, column_sort=None):
+        """
+        Adapter for Pandas DataFrames in long format, where different time series for the same id are
+        labeled by column `column_kind`.
+
+        :param df: the data frame
+        :type df: pd.DataFrame
+
+        :param column_id: the name of the column containing time series group ids
+        :type column_id: str
+
+        :param column_kind: the name of the column containing time series kinds for each id
+        :type column_kind: str
+
+        :param column_value: the name of the column containing time series values
+        :type column_value: str
+
+        :param column_sort: the name of the column to sort on
+        :type column_sort: str|None
+        """
+
+        _check_nan(df, column_id, column_kind, column_value)
+        _check_colname(column_kind)
+
+        self.column_value = column_value
+
+        if column_sort is not None:
+            _check_nan(df, column_sort)
+            self.df_grouped = df.sort_values([column_sort]).groupby([column_id, column_kind])
+        else:
+            self.df_grouped = df.groupby([column_id, column_kind])
+
+    def __len__(self):
+        return len(self.df_grouped)
+
+    def __iter__(self):
+        for group_key, group in self.df_grouped:
+            yield Timeseries(group_key[0], str(group_key[1]), group[self.column_value])
+
+
+class TsDictAdapter(TsData):
+    def __init__(self, ts_dict, column_id, column_value, column_sort=None):
+        """
+        Adapter for a dict, which maps different time series kinds to Pandas DataFrames.
+
+        :param ts_dict: a dict of data frames
+        :type ts_dict: dict[str, pd.DataFrame]
+
+        :param column_id: the name of the column containing time series group ids
+        :type column_id: str
+
+        :param column_value: the name of the column containing time series values
+        :type column_value: str
+
+        :param column_sort: the name of the column to sort on
+        :type column_sort: str|None
+        """
+
+        _check_colname(*list(ts_dict.keys()))
+        for df in ts_dict.values():
+            _check_nan(df, column_id, column_value)
+
+        self.column_value = column_value
+
+        if column_sort is not None:
+            for key, df in ts_dict.items():
+                _check_nan(df, column_sort)
+
+            self.grouped_dict = {key: df.sort_values([column_sort]).groupby(column_id)
```

Comment:
Just out of curiosity, how expensive is that groupby? Is it saving the groups in memory?

Reply:
Thanks for reviewing.

Regarding your first question, yes one could investigate if long and dict could be iterated more efficiently. But don't think this is critical. To your second question, I must say that I am relatively new to pandas and the way it behaves is a bit fuzzy to me. And to most other people, if I look at SO. For the case of groupby, I found this: https://stackoverflow.com/questions/52711823/does-groupby-in-pandas-create-a-copy-of-the-data-or-just-a-view

Reply:
yeah, the internal workings of pandas like the blockmanager of ((https://uwekorn.com/2020/05/24/the-one-pandas-internal.html)) are also fuzzy to me